compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -521,6 +521,76 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
   return success();
 }
 
+/// Returns true if the operation is a GenericOp implementing a 2D transpose.
+static bool isTransposeOp(linalg::LinalgOp linalgOp) {
+  if (!isa<linalg::GenericOp>(linalgOp)) return false;
+  // Check that the op has 2 parallel loops.
+  if (linalgOp.getNumParallelLoops() != 2) {
+    return false;
+  }
+
+  // Check that all the iterators are parallel.
+  if (linalgOp.getNumParallelLoops() != linalgOp.getNumLoops()) {
+    return false;
+  }
+
+  // Check that the op has only one input and one output.
+  if ((linalgOp.getNumInputs() != 1) || (linalgOp.getNumOutputs() != 1)) {
+    return false;
+  }
+  // Check for 2D operations
+  auto inputShape =
+      linalgOp.inputs()[0].getType().cast<ShapedType>().getShape();
+  auto outputShape =
+      linalgOp.outputs()[0].getType().cast<ShapedType>().getShape();
+  if (inputShape.size() != 2 || outputShape.size() != 2) {
+    return false;
+  }
+
+  // Only transpose static sizes
+  if (inputShape[0] == ShapedType::kDynamicSize ||
+      inputShape[1] == ShapedType::kDynamicSize ||
+      outputShape[0] == ShapedType::kDynamicSize ||
+      outputShape[1] == ShapedType::kDynamicSize) {
+    return false;
+  }
+
+  // Check that the two indexing maps are a permutation of each other.
+  auto indexing_maps = linalgOp.getIndexingMapsArray();
+  return !indexing_maps[0].isEmpty() && !indexing_maps[1].isEmpty() &&
+         ((indexing_maps[0].isIdentity() && !indexing_maps[1].isIdentity() &&
+           indexing_maps[1].isPermutation()) ||
+          (!indexing_maps[0].isIdentity() && indexing_maps[0].isPermutation() &&
+           indexing_maps[1].isIdentity()));
```

Comment:
Here are some ideas on how to match affine maps:
https://github.com/llvm/llvm-project/blob/86b67a310dedf4d0c6a5bc012d8bee7dbac1d2ad/mlir/lib/Conversion/VectorToGPU/VectorToGPU.cpp#L302

I think you want to start by matching only this case: {(d0, d1) -> (d1, d0), (d0, d1) -> (d0, d1)}

Reply:
I got this specific permutation check from https://github.com/iree-org/iree/blob/658d3edbe76527c02596643bbb5990807c0028cd/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L976.

 I believe this code already does exactly what you're suggesting, but also handles the reverse case. Its also general to accept any transpose permutation (for higher ranked tensors). 

looking at the code you linked, would like to discuss further exactly what its doing